### PR TITLE
MariaDB Database Backend

### PR DESCRIPTION
**Pull request by caffeinatedblocks**
Adds MariaDB support for map, auth, player, and mod storage backends.

Fully tested and ready to build/play. This PR is ready for review.

You will need to [install MariaDB Connector C++](https://mariadb.com/docs/skysql/connect/programming-languages/cpp/install/), then simply add "-DENABLE_MARIADB=ON" and build.

Oh. P.S. Note that mod storage is "less capable" than SQLite/PostgreSQL: the mod name and key are both limited to 128 chars. The primary key, for whatever reason, was designed to use a combined index of (mod_name, key), so these values really should not be "unlimited" as they are with SQLite/PostgreSQL. The values I chose could be increased by modifying database-mariadb.cpp at lines 1261 and 1262, respectively, however I was unable to determine if there is any actual need / use case for this. MariaDB offers many storage engines, and each may vary. InnoDB has a maximum 1536 or 768 byte index (depending on page size). I chose 128 as a safe upper limit to accommodate any storage engine, so this value may need to be increased if you have mods which use larger names / keys. If there is an actual need, I would be happy to come up with a better way to handle mod storage in a self-contained fashion which would not affect any other backends.

### CODE REVIEW

**Comment by sfan5** [comment]
I once wrote a MySQL database backend (map-only, was years ago) but decided not to open a PR because I couldn't come up with any good reasons to use it when Postgres already existed. Do you have any in mind?

**Comment by caffeinatedblocks** [comment]
Without sounding argumentative? That might prove difficult.

1. I don't know PostgreSQL.
2. I know MySQL/MariaDB very well.
3. I'm sure I'm not the only one who identifies with points 1 and 2.
4. I've spent over 20 years working with MySQL.
5. The MariaDB team is very dedicated and eager to help.
6. I had a chat with MariaDB's founder/CTO and current lead developer (also co-founder of MySQL), and upon explaining our use case, and demonstrating SQLite and PostgreSQL's advantages over MariaDB in terms of insert query efficiency, he was happy to open an issue with highest priority to help make sure MariaDB becomes more performant to meet the needs of our project. See: https://jira.mariadb.org/browse/MDEV-30736
7. MySQL/MariaDB has countless open-source tools to help manage, monitor, backup, export, and otherwise maintain the databases.
8. I spent a month of my life researching, studying, learning c++, and talking with MariaDB devs to make this happen.
9. It works, and offers users another great choice for backends.
10. MariaDB, unlike SQLite and Postgre, can be configured with a wide variety of storage engines, and even run totally in-memory like Redis.
11. It's extremely performant and reliable (not that the others aren't.
12. Why not?

File: src/database/database-mariadb.cpp
Comment (nerzhul) [inline src/database/database-mariadb.cpp:52]: you can optimize a bit the code by not putting a `;` on each line ending and calling oss only on first line

File: src/database/database-mariadb.cpp
Comment (nerzhul) [inline src/database/database-mariadb.cpp:354]: never do a SELECT *, always use precise fields

File: src/database/database-mariadb.cpp
Comment (nerzhul) [inline src/database/database-mariadb.cpp:538]: put TEXT for attr too

File: src/database/database-mariadb.cpp
Comment (nerzhul) [inline src/database/database-mariadb.cpp:504]: i'm not sure about the VARCHAR(60), i think you should use TEXT like on PG to prevent bad silent truncates

File: src/database/database-mariadb.cpp
Comment (nerzhul) [inline src/database/database-mariadb.cpp:587]: just SELECT name, not all fields, this will optimize fetch time DB side

**Review by nerzhul — CHANGES_REQUESTED** [review]

File: src/database/database-mariadb.cpp
Comment (rubenwardy) [reply src/database/database-mariadb.cpp:504]: Player username absolutely does not need to be unlimited, I believe it's already limited to something like 18-20 characters

Comment (nerzhul) [reply src/database/database-mariadb.cpp:504]: @rubenwardy do you know that mysql will just truncate field silently and it's  not a constraint ?

Comment (rubenwardy) [reply src/database/database-mariadb.cpp:504]: If that's the cast, then it should be enforced by the server at an earlier point

Comment (nerzhul) [reply src/database/database-mariadb.cpp:504]: it should, varchar is just a text with limitation on storage size, there is zero interest db side. in terms of index it's same as text, then no interest too :)

Comment (rubenwardy) [reply src/database/database-mariadb.cpp:504]: The max player name size is 20, as defined in player.h. Attempts to connect with a name longer than this is rejected.

#define PLAYERNAME_SIZE 20

Comment (rubenwardy) [reply src/database/database-mariadb.cpp:504]: > it should, varchar is just a text with limitation on storage size, there is zero interest db side. in terms of index it's same as text, then no interest too :)

Please rephrase because I have no idea what you mean

Comment (rubenwardy) [reply src/database/database-mariadb.cpp:504]: I'm guessing you're referring to the fact that the performance of TEXT and VARCHAR are the same in PostgreSQL:

> Both TEXT and VARCHAR have the upper limit at 1 Gb, and there is no performance difference among them (according to the PostgreSQL documentation).

This is not the case for MariaDB, where VARCHAR is faster. Additionally, using VARCHAR with postgresql still has the benefit of limiting the size of stored data

**Comment by rubenwardy** [comment]
I assume by @nerzhul's review that he supports this PR

**Comment by celeron55** [comment]
It may be unwise to add MariaDB support as it will have to be then maintained essentially forever by the core team while it doesn't really bring anything to the table compared to PostgreSQL.

It would make sense to at least have the community vote about this so that we know whether there's actual need for this.

**Comment by rollerozxa** [comment]
I've always wanted a MariaDB database backend for Minetest, but never mentioned it because I assumed it would be shut down by Postgre elitists anyways, and didn't feel capable enough to implement it myself. Just because nobody has had the courage to open a feature request for it doesn't mean it isn't wanted by people. FWIW, there has been a couple of mentions of it in the past by others both on Discord and in IRC.

MariaDB support is good if you're hosting a Minetest server that coexists with a LEMP (or similar) stack on the same machine. Which is where I'm coming from, I'd like to piggyback off an existing MariaDB server and get the same benefits the Postgre backend has without having to maintain another database server on the same (rather low-end) VPS.

**Comment by rollerozxa** [comment]
That being said though, I can't even get it to compile with the MariaDB library installed right now, so there's some things that need to be sorted out first before it can be merged.

**Comment by lhofhansl** [comment]
+1 from me. I will not be using MariaDB, but I can see how that will be useful for many folks.
In terms of maintenance... I do not see many changes related to Postgres or SQLite.

Nit: It is either "Postgres" or "PostgreSQL" (where is s is cleverly pulled into SQL). It's *never* Postgre.

File: src/database/database-mariadb.cpp
Comment (x2048) [reply src/database/database-mariadb.cpp:52]: Example: https://linuxhint.com/c-multiline-string-literals/

File: src/database/database-mariadb.cpp
Comment (x2048) [inline src/database/database-mariadb.cpp:464]: Output block position to the errorstream, so server admin can act on it.

File: src/database/database-mariadb.cpp
Comment (x2048) [reply src/database/database-mariadb.cpp:354]: Looking at loadBlock below, only `data` is needed

Comment (appgurueu) [reply src/database/database-mariadb.cpp:354]: > never do a SELECT *, always use precise fields

Why? To make changing the schema less likely to have side effects?

File: src/database/database-mariadb.cpp
Comment (caffeinatedblocks) [reply src/database/database-mariadb.cpp:52]: I didn't know the best way to handle multiple lines and throughout the engine code, the style varies, so I went iwth what was done in src/server.cpp at startup with the ASCII art and following lines. In any case, pushing new commit now to resolve.

File: src/database/database-mariadb.cpp
Comment (appgurueu) [reply src/database/database-mariadb.cpp:587]: I doubt it. Keep in mind this is an `EXISTS` statement. The optimizer should definitely be smart enough to optimize this out. I see no reason to add verbosity here.

File: src/database/database-mariadb.cpp
Comment (x2048) [inline src/database/database-mariadb.cpp:437]: This makes two copies of the block stream, one in data and one in *block.

File: src/database/database-mariadb.cpp
Comment (x2048) [inline src/database/database-mariadb.cpp:471]: What is the point of inverting endianness of the positions?

File: src/database/database-mariadb.cpp
Comment (caffeinatedblocks) [reply src/database/database-mariadb.cpp]: Fixed in latest commit.

File: src/database/database-mariadb.cpp
Comment (caffeinatedblocks) [reply src/database/database-mariadb.cpp]: Fixed in latest commit. I made attr a VARCHAR because primary key is (player, attr) in Postgre implementation, however upon review, I see that this was a mistake because all player_metadata queries only use name, so Postgre's index is screwed up.

File: src/database/database-mariadb.cpp
Comment (caffeinatedblocks) [reply src/database/database-mariadb.cpp]: Fixed in latest commit. Now importing and using PLAYERNAME_SIZE from player.h.

File: src/database/database-mariadb.cpp
Comment (caffeinatedblocks) [reply src/database/database-mariadb.cpp]: Fixed in latest commit.

File: src/database/database-mariadb.cpp
Comment (caffeinatedblocks) [reply src/database/database-mariadb.cpp]: Fixed in latest commit.

File: src/database/database-mariadb.cpp
Comment (caffeinatedblocks) [reply src/database/database-mariadb.cpp]: Fixed in latest commit.

File: src/database/database-mariadb.cpp
Comment (caffeinatedblocks) [reply src/database/database-mariadb.cpp]: Great point. I had that from when I was first beginning, and using Postgre implementation as a reference. I later realized that wasn't necessary with other queries... Fixed in latest commit.

**Comment by caffeinatedblocks** [comment]
All suggestions, corrections, etc. have been accounted for, and additional optimizations with regard to queries and table definitions have been made.

**Comment by caffeinatedblocks** [comment]
> It may be unwise to add MariaDB support as it will have to be then maintained essentially forever by the core team while it doesn't really bring anything to the table compared to PostgreSQL.
> 
> It would make sense to at least have the community vote about this so that we know whether there's actual need for this.
> 
> EDIT: It is notable that there was no issue requesting MariaDB support.

It does bring something to the table that Postgre doesn't: MySQL/MariaDB support. I've already covered a number of benefits, but the most important is that more hosting environments include MySQL/MariaDB by default than any other RDBMS.

Anyway, that aside, all the more reason to let me join the team! I'll be happy to continue supporting it and continuing development as MariaDB receives more tweaks and optimizations in the future. 😁

**Comment by celeron55** [comment]
Just to be clear: I'm not blocking this PR. I merely pointed out some things. Feel free to continue with your own judgement.

**Comment by caffeinatedblocks** [comment]
> Just to be clear: I'm not blocking this PR. I merely pointed out some things. Feel free to continue with your own judgement.

I get it, but coming from you, the statement "it doesn't really bring anything to the table" carries a lot of weight. My concern would be that community members (especially those who don't know better) would blindly trust and adopt your opinion as their own based solely on your own personal merits, rather than the validity of your statement, e.g. "He said so, so it must be true."

I think any healthy, responsible open-source community should take care not make such bold statements in such a public setting. To say that the software which continues to hold the vast majority of market share decade after decade brings nothing to the table in comparison to its underlings seems a little naive to me.

**Comment by joepie91** [comment]
(Note that I am posting this as someone just passing through; I am not otherwise involved in Minetest.)

> I've already covered a number of benefits, but the most important is that more hosting environments include MySQL/MariaDB by default than any other RDBMS.

This seems like a particularly strange argument to me. While I could understand this argument in the context of, say, a PHP project... Minetest *already* is essentially custom software that can only run within a hosting environment that allows non-provider-supplied daemons to be run. Such an environment would automatically also make it possible to run PostgreSQL.

What sort of environment are you thinking of, where installing Minetest is no problem, but installing PostgreSQL is?

---

More broadly: over the years, I've seen many, *many* projects attempt to support 'swappable database backends', usually motivated by some sort of argument about user choice or "best fit for individual users' usecases". Without exception, they ran into two major problems:

1. It turns out that despite many different RDBMSes being nominally "SQL databases", the *actual* shared feature and syntax set between them is extremely limited. Many things have changed since the introduction of the SQL standard, and in practice, basically nobody actually uses any RDBMS strictly according to the SQL standard. This means that maintaining multiple such backends *reliably* ends up requiring more maintenance effort than anticipated upfront, to deal with database-specific oddities. This can make sense for SQLite, to have a self-contained option; but for two databases that follow fundamentally the same operational model (like MySQL and PostgreSQL), it's not so clear.
2. The original goal is usually something along the lines of "so people can pick the option that works best for their case", but... because of the practical complexities of maintaining support for different flavours of SQL, the only viable solution generally ends up being "using only a lowest-common-denominator subset of features that works in all backends", which not only makes it difficult to take advantage of database-specific features and optimizations, but also means that the whole point of swappable backends becomes moot - if you use the same subset of features in both anyway, why would you need to swap one for the other in the first place?'

These issues are not hypothetical - there are significant differences between PostgreSQL and MySQL/MariaDB even in just how TEXT fields are handled, for example, that have far-reaching effects on the database schema and the (performance) limitations imposed by it.

IMO, to justify asking the maintainers to maintain an additional database backend - which is the practical outcome of "adding a backend" - the new backend would need to concretely solve a problem that people demonstrably have with the current setup, not just have an impressive list of features and benefits that could go on a marketing slide. I'm personally not convinced that that is the case here.

**Comment by Montandalar** [comment]
Despite various comments that cast doubt on this pull requests based on maintenance burden and question its necessity, we should instead consider that there are already a number of database backends and that their maintenance does not seem by the evidence to have been a burden.

Redis support for worlds was added in 2014. There has barely been a mention of it in the git logs since, with only 18 commits touching the relevant file, and no commits since 2017. Unless someone has recent experience with redis and can vouch that it has problems, I would be more inclined to conclude Minetest's support for it is stable and mature. Again, I am no expert, but even with the major change from Redis 5 to Redis 6, [migration seems easy enough](https://docs.redis.com/latest/rs/installing-upgrading/upgrading/upgrade-cluster/).

LevelDB similarly was added back in 2012, cpp file only touched in 18 commits, last touched in 2017, and that last only in the commit 29ab20c that was not database-specific. I am not experienced with LevelDB either, but LevelDB has not seen many major updates between 1.20 and 1.23, and 1.23 was back in 2021. Both LevelDB and Minetest's support for it seem mature, with the former exception of Android, where it was removed as an option.

This Android exception for LevelDB I would say proves the rule: We do not have to accept the burden of supporting database backends and platforms where we do not have the expertise to do so. I haven't investigated in depth, but  I think LevelDB support for Android was lacking in upstream's mainline. Unlike LevelDB, I think given the prevalence of MariaDB and the efforts of this author, I'd say that we do in fact have the expertise to support it properly.

As to the next objection: That swappable database backends share only a small set of common functionality, and the end result must be only a least common denominator, and cannot take advantage of database-specific optimisations. May I ask what exact optimisations you think are important for Minetest specifically? And why would swappable database support preclude those, instead of *enabling* specialisations and optimisations? Minetest is not doing much, as far as I'm aware, beyond basic CRUD on mapblocks and needing a good index. Finally, can you cite any of your claims in writing and external links? I am not terribly convinced about database expertise from someone whose Gitea seems to be filled with JavaScript projects. I am not saying that you don't potentially have private code somewhere else, or colleagues and collaborators who can back you up but: please enlighten us with some blog posts, pull request threads, and so on.

And finally I will say: If we were to object to any new development in Minetest on the basis of it being unnecessary, then we could close up all feature development with the exception of bug fixes right now. In fact, with such a Google-type philosophy, one could even say it would be smarter to start to drop database backends. However, Minetest as a free software project has, I believe, a social responsibility not to drop such support. Nor should Minetest try to be minimalist when it comes to features. Past backwards-incompatible changes to Minetest, even minor ones, have been met with resistance, such as Minetest 5.0's breaking changes and the attempted dropping of TGA support. There is no need to try to put more fuel on the fire. One could argue that there are better "actually-a-new-feature" items to put on the agenda - but when the code is already offered by a new contributor, I don't see that there is much to lose by spending the time to review it

The above comment by joepie91 mostly spreads uncertainty and doubt about the usefulness of MariaDB, but I like to be more optimistic and welcoming of new ideas and new contributors. Having MariaDB as an option creates more opportunities to leverage expertise from MariaDB experts who want to easily run a Minetest server and create server management tools. Minetest has a historical precedent of adding new database backends, not refusing them due to objections that pop up. I'd say give it a fair code review and add it if the code is quality and it works after some playtesting.

**Comment by joepie91** [comment]
> Finally, can you cite any of your claims in writing and external links? I am not terribly convinced about database expertise from someone whose Gitea seems to be filled with JavaScript projects. I am not saying that you don't potentially have private code somewhere else, or colleagues and collaborators who can back you up but: please enlighten us with some blog posts, pull request threads, and so on.

If this commentary is representative for how project contributors interact with folks, I frankly think you'll have bigger problems where "welcoming new contributors" is concerned.

I don't really have any interest in engaging with the rest of your post, given the hostility therein. As I mentioned, I'm commenting as a passer-by on a problem that's often poorly-understood and therefore could use some explanation. If the maintainers of Minetest feel differently about the tradeoffs, then that's fine too - it is not my call to make.

**Comment by appgurueu** [comment]
I still don't see a good enough case having been made for supporting this backend and reviewing a 2kloc pull request. To the central question of what the use case of this is and what advantages it brings over the existing options posed by sfan5:

> I once wrote a MySQL database backend (map-only, was years ago) but decided not to open a PR because I couldn't come up with any good reasons to use it when Postgres already existed. Do you have any in mind?

caffeinatedblocks replied with [the following list](https://github.com/minetest/minetest/pull/13266#issuecomment-1451153991), where the only "real" points except for "I and others are already invested in MariaDB (and this PR)" are

* "MariaDB, unlike SQLite and Postgre, can be configured with a wide variety of storage engines, and even run totally in-memory like Redis." (but we do already have Redis, so if you want to run fully in-memory why not use that?)
* "Oh, also, it opens doors because most every web host on the planet provides MySQL/MariaDB, whereas SQLite and Postgre are often not (or require additional installation, etc.)" (refuted by @joepie91)

So really, I fail to see the use case of "MariaDB works for me but Postgres doesn't" beyond "I'm somehow invested in MariaDB".

I'll play the devil's advocate of the "Postgres elitist" here: Is there anything MariaDB, or this backend, does fundamentally better than Postgres? All I see is a fresh arbitrary restriction of mod name and key to 128 chars.

I wouldn't outright reject this PR, but I don't see why it shouldn't be low priority considering the ratio of added value to review effort.

And if you argue that the review effort wouldn't be significant since "it's just adapted copypaste from the Postgres backend anyways" (or similar), then a code deduplication - perhaps among all SQL backends - would be in order.

**Comment by sfan5** [comment]
I think the arguments were reasonably explained in a way that was understandable to anyone with a medium amount of database experience and I'm inclined to agree that the mentioned risks exists. In that context saying "what are your credentials, I only see you doing JS" (paraphrased) is not very nice.
That said, I don't think they apply to Minetest. We use databases as little more than key-value stores, we don't have complicated SELECT queries or complex relationships.

I plan to review this PR and wouldn't mind it getting in (though I agree with @appgurueu that it's low priority).

---

What I'd rather see discussed is this point I made: https://github.com/minetest/minetest/pull/13266#issuecomment-1454033861
I see two options:
1. Can we reasonably vendor mariadb-connector-cpp in `lib`? How many LOC is it?
2. If not, how about rewriting the code to work with mariadb-connector-c?

Left as-is it would be sad for usability since mysql-connector-cpp only seems to be [packaged](https://pkgs.org/search/?q=mysql-connector-c%2B%2B) by two BSD flavors and EPEL.

**Comment by Sur3** [comment]
Stupid question, but why do PostgresSQL and MariaDB need different backends, as they both talk SQL there could be just one general SQL backend for all SQL-type databases with IP:PORT connection, no?!

**Pull request closed without merging**
